### ISSUE TITLE
Blackjack environment does not handle naturals correctly

### ISSUE THREAD

**Issue by rhalbersma** [openai/gym#1879]
**TL;DR**: the current blackjack code does not correctly implement the Sutton and Barto example, which in turn also does not correctly implement the usual casino rules as described by Thorp, "Beat the Dealer".

Lines 105-107 of `blackjack.py` do not correctly implement the scoring of naturals (21 with 2 cards). A natural by either the player or the dealer beats a 21 from more than 2 cards from the other player. The current code rewards this with 0. 

A fix is to select the score from a small dictionary, keyed on the combination of `is_natural` on the player's and dealer's hands. Since the case of the player busting has already been handled, it is also convenient to directly test for the dealer busting, which simplifies the scoring of regular hands to `sum_hand`, obviating the need for `score()`.

```
if is_bust(self.dealer):
    reward = 1.
else:
    reward = {
        (True, False)   : 1.5 if self.natural else 1.,
        (True, True)    : 0.,
        (False, True)   : -1.,
        (False, False)  : cmp(sum_hand(self.player), sum_hand(self.dealer))
    }[(is_natural(self.player), is_natural(self.dealer))]
```

Note that Sutton and Barto correctly describe the positive reward for a player natural versus a dealer 21 (albeit that they use +1 instead of +1.5 as described by their cited Thorp). However, also Sutton and Barto neglect to mention the case of a player 21 against a dealer natural, scoring it as 0 instead of -1. This might also explain their discrepancy (mentioned on p.99) with Thorp: S & B recommend hitting on a soft 18 with a dealer's ace showing. Since this ace gives the dealer a 4/13 probability of a natural, even a player's 21 is will not help him some of the time. This will make it better to stand on such a soft 18.

I can submit a PR for a 'Blackjack-v1' (since fixing this scoring is a breaking change). If such a PR is welcomed, I'd also want to fix some other small things, such as the order of the returned observation. It's more convenient to return this a `(usable_ace, player count, dealer card)` and to convert `usable_ace` to an `int` instead of a `Boolean`. The integer conversion makes it possible to directly index NumPy arrays with the returned observation tuple, which simplifies table-based algorithms. The permuted order of `usable_ace` makes displaying NumPy arrays (for Q-values / policies e.g.) naturally appear in the format of S & B, i.e. two matrices, one for soft and hard hands each.

**Comment by christopherhesse** [comment]
Hi thanks for catching this!  Since this is meant to be an implementation of the Sutton and Barto environment, I think we are only interested in differences between what is in that book and what is implemented in the code.

Incrementing the environment number would be the thing to do here, and it would be extra nice if `Blackjack-v0` remained in the code base.  This may make it not worthwhile to modify other small things since both old and new versions would have to be supported.  We would welcome a PR for the bugfix though, especially if it kept around support for `Blackjack-v0`

**Comment by jkterry1** [comment]
@rhalbersma so I finally got to independently review this, and I think you're correct that this is a real issue. Are you still interested in doing the upgrade for this?

**Comment by RedTachyon** [comment]
So I started looking at this in more detail, and it's a bit of a mess.

First we have the Thorp/casino rules which I'm frankly not familiar with. Let's say this is irrelevant - we want consistency with S&B

According to Example 5.1, rewards are only +1, 0, -1, for winning, drawing and losing respectively. There's no special reward for a natural. This is not valid at the moment, since in case of a natural, the reward is +1.5
It is mentioned that a natural by the player automatically beats a non-natural by the dealer. BUT there is nothing about a dealer's natural. The only thing is that if both the player and the dealer have naturals, it's a draw. 

So reading the text literally, (dealer natural 21) vs (player non-natural 21) should result in a draw, because it falls back to the default rules. This is current behavior, but opposed to what OP said.

To make things more interesting, `Blackjack-v0` doesn't actually have this issue, since the `natural` argument takes the default value of `False`, which means that naturals are completely ignored, as if on purpose.

My suggestion is to keep `Blackjack-v0` as is (obviously), and make a Blackjack-v1 which implements S&B as closely as possible. This would mean having rewards always be in the set {-1, 0, 1}, and having naturals autowinning only for the player (and drawing when the dealer also has a natural)

PR incoming

I don't think it's very urgent, so if someone (@rhalbersma) can point out why a dealer natural should win over a player non-natural, that'd be great and to be fixed afterwards.

**Comment by jkterry1** [comment]
"We want consistency with S&B"
Yes, exactly. If you want casino style, use RLcard.

"My suggestion is to keep Blackjack-v0 as is (obviously)"
Why is this desirable? It seems like it should be removed in favor of a deprecation ​warning like any old environment would be.


Also, can you please explain the natural issue in words simple enough for me to understand?

**Comment by jkterry1** [comment]
Additionally, Gym being what it is, asking Sutton's group for clarification is a legitimate option for us if required.

**Comment by RedTachyon** [comment]
"Why is this desirable? It seems like it should be removed in favor of a deprecation ​warning like any old environment would be."

That's fair, I guess I got influenced by the previous discussion. I think keeping the current code supporting both versions isn't a problem (if someone needs a different version, they can set the arguments manually, and it's a 2-line difference), but the registration can be removed.

---

So the thing with naturals:
In Blackjack, it might happen that the first two cards of either the player or the dealer, sum up to 21. This is known as a natural.

There's the question of what to do in the case of a natural (on either side)

According to my understanding of S&B: player's natural wins with everything except a dealer's natural - then it's a draw. Otherwise, follow normal rules.

According to what OP said - what I said above, but also a dealer's natural wins with everything except a player's natural - then it's a draw.

The difference is the situation where the dealer gets a natural, and the player has e.g. 3 cards that sum up to 21. According to OP (and apparently casino rules?), this is a win for the dealer. According to my literal reading of S&B, this is a draw.

Worth noting that the current behavior is wrong from both perspectives, because it doesn't do "player's natural wins with everything".

---

Then there's the question of rewards. In the current implementation, you get -1 for a loss, 0 for a draw, 1 for a regular win, and 1.5 for a natural win if naturals are turned on. It's worth mentioning that in Blackjack-v0, naturals are turned off, but this can be "circumvented" by manually calling the environment class with `natural=True`.

According to S&B, the rewards are -1, 0, 1 as described earlier, and there's no mention of 1.5.

So while OP kept it at 1.5, I changed it to 1.0 for full consistency.

Hope that's clear

---

And asking Sutton's group could definitely help, I'm all for that

**Comment by jkterry1** [comment]
Thank you for the detailed explanation. All this sounds good, two notes:

- If the natural rule you described unreasonable rule from S&B reaching out is within the question
- Whatever the default rules are they need to be the S&B are they need to be in the book; can you confirm that's the case?

**Comment by RedTachyon** [comment]
I'm honestly not familiar enough with casino rules to tell whether or not what's described in the book is unreasonable.

In any case, my intention is indeed that the default initialization of Blackjack-v1 matches the behavior described in the book, interpreted as literally as possible, if that is what you mean. 

"If the player has 21 immediately (an ace and a 10-card), it is called a natural. He then wins unless the dealer also has a natural, in which case the game is a draw. If the player does not have a natural, then he can request additional cards [...usual game rules]"

Naturals are not mentioned in the example afterwards, so there is no mention of a situation where the dealer has a natural, and the player doesn't.
